### ISSUE TITLE
Suggestions to enrich the Provider API

### ISSUE THREAD

**Issue by bsautel** [gradle/gradle#12974]
I recently refactored the [gradle-node-plugin](https://github.com/node-gradle/gradle-node-plugin) to use lazy configuration by using the `Provider` and `Property` concept as recommended in the Gradle documentation. I encountered two limitations with the API that I would like to share with you.

1. The plugin has [an extension with multiple properties](https://github.com/node-gradle/gradle-node-plugin/blob/5bd125e1605a47e456c78734f6c5641357fbe0fc/src/main/kotlin/com/github/gradle/node/NodeExtension.kt) and most of the tasks also have several properties. I tried not to break (by calling `.get()` for instance) the `Provider` chain. For that, we have to use mapping operations as described in the docs. But most of the time, to compute the value of a property, I needed multiple properties (mapping multiple properties to one). None of the `map` and the `flatMap` enabled me to handle simply this use case.

I wrote some helper methods that are able to map multiple properties to one inspired by the [ReactiveX zip operator](http://reactivex.io/documentation/operators/zip.html). My helper methods are visible [here](https://github.com/node-gradle/gradle-node-plugin/blob/2fbcfdf71b553cd3aadbbcdb5104a7d9f59f8b89/src/main/kotlin/com/github/gradle/node/util/Provider.kt). I think that it would be nice to integrate that directly in the Gradle API, that would avoid plugin authors to write this themselves.

This enables to write something like this:
```kotlin
val aProvider: Provider<Int>
val bProvider: Provider<Int>
val sumProvider = zip(aProvider, bProvider).map { (a, b) -> a + b }
```

2. When a property (or a provider) is not defined, it contains no value and `getOrNull()` returns `null`. As far as I could see, when the provider contains no value, the `map` never calls the mapping function, it directly returns a provider with no value. But in some cases, I would like to do some mapping operations even if the value is `null`, like `getOrNull()` does, but without breaking the mapping chain.

Let's take an example of something that I could not do. I would like to declare a task output only if the `optionalProperty` is not defined. This means I want a `Provider` containing a value if `optionalProperty` is empty and containing no value if `optionalProperty` is not empty.

In this example I used the `flatMapEvenIfNull` method that does not exist but that would solve my issue. It would have the same behavior as `flatMap` excepted that it would also map with the value `null` when the `Provider` contains no value. If I use `flatMap`, the resulting `Provider`will be always empty.
```kotlin
val optionalProperty: Provider<String>
val directory: Provider<Directory>
@get:OutputDirectory
@get:Optional
val taskOutput: Provider<Directory> = optionalProperty.flatMapEvenIfNull {
    if (it == null) directory else project.providers.provider { null }
}
```
I tried to do that using `orElse` but as far as I know it does not enable me to do that.

For now, I had to break the `Property` chain by calling `getOrElse()` to do that.

Maybe the Gradle SDK already contains some tools to do that. I spent a lot of time looking for a solution but maybe I just did not find it. In this case, please let me know and excuse me.

Hope this will help to improve Gradle which is already a wonderful build tool!

**Comment by big-guy** [comment]
+1 for the zip idea.

Do you have an example of where you need/are doing the second suggestion (`flatMapEvenIfNull`) in the plugin?

**Comment by bsautel** [comment]
Yes, you can see it [here](https://github.com/node-gradle/gradle-node-plugin/blob/d7c93365f4a257f3b6968dd1a80a05d3280814be/src/main/kotlin/com/github/gradle/node/npm/task/NpmInstallTask.kt#L68). Let me give you some context.

This is the `npm install` task which creates or updates the `node_modules` directory that contains the project dependencies. Oh, that's simple, let's declare this directory as a task output!

Yes, but unfortunately it is not so simple. Some users reported us that there are some specific cases (not really legal I think but they exist) in which the code contained in the dependency modifies itself in the package directory (node-gradle/gradle-node-plugin#38). In the Java world, we could compare that to a jar that you download from a repository which, when you execute it, modifies itself. 🤔

The problem with that behavior is that the `npm install` task is no longer up-to-date and always runs (and it is not very fast). When it runs, it restores the dependency to its original content and when the dependency code runs, it modifies it again.

To workaround this issue, we added an option that enables to filter out some patterns from the `node_modules` directory to target specifically those files that are updated. This is what does the `nodeModulesOutputFilter` option of the task. In this case, we use a file tree since we cannot filter out a directory.

After releasing this change, we were reported that the `npm install` task was much slower than before, and specifically its up-to-date check (node-gradle/gradle-node-plugin#63). It sounds like `@OutputFiles` on a `FileTree` is slower than `@OutputDirectory` on a `File` (it's not so surprising).

To fix this issue we had to declare two optional outputs and exactly one of them is not `null` according to the configuration. If you define a filter, it is the `@OutputFiles`, else it is the `@OutputDirectory`.
So, for the `@OutputDirectory`, we want it to be defined only when the `nodeModulesOutputFilter` is not defined. This is here I would like to `map` (rather `flatMap`) even if it's `null`. I did not find any other way than breaking the `Provider` chain using `getOrNull()`.

Hope my explanation was clear!

And maybe it exists another way to handle this use case that would be more relevant, in this case, let me know!

**Comment by big-guy** [comment]
> To fix this issue we had to declare two optional outputs and exactly one of them is not null according to the configuration. If you define a filter, it is the @OutputFiles, else it is the @OutputDirectory.
> So, for the @OutputDirectory, we want it to be defined only when the nodeModulesOutputFilter is not defined. This is here I would like to map (rather flatMap) even if it's null. I did not find any other way than breaking the Provider chain using getOrNull().

I'd have to play around with it a bit to see if there's a difference between `@OutputDirectory` and `@OutputFiles` here, but I think you could potentially do something like this (pseudocode):

```
class NpmInstallTask {
    // This specifies the directory to install things into.
    // This is turned into a file path and passed to npm
    @Internal
    DirectoryProperty npmInstallDirectory

    // This is the output of the task that's consider for up-to-date checking
    @OutputFiles
    ConfigurableFileCollection npmInstalledFiles
}

// In NpmPlugin

npmInstall.npmInstallDirectory.convention( /* wire to node extension */)
npmInstall.npmInstalledFiles.from(npmInstall.npmInstallDirectory)
```

And then someone who needs to filter things from the install directory would do (this could be put behind a method):
```
npmInstall.npmInstalledFiles.from = fileTree(npmInstall.npmInstallDirectory, configurableFileTree -> {
    // filter configuration
})
```

There are a few general improvements I think you could do to the plugin:
- Depending on how old of a version of Gradle you'd like to support, you could take advantage of [managed properties](https://docs.gradle.org/current/userguide/custom_gradle_types.html#managed_properties) to reduce boilerplate when using `Property` and `Provider` types. This works for tasks and extensions.
- I would recommend trying to make the tasks as "dumb" as possible and keep as much configuration of the task in the plugin. This means that it's OK if the task uses `DirectoryProperty` instead of `Provider<Directory>` so you can set it from the plugin. I think this might clean up some of the "by lazy" properties you have.
- I think you don't need to explicitly finalize the values in the extension. This happens when the tasks are put into the dag
- If you can, avoid calling `getProject()` in your task (APIs like copy, exec, files, etc have equivalents in more focused services that you can get).  This doesn't change anything yet, but we'll eventually start making these kinds of APIs more restrictive in tasks.

**Comment by bsautel** [comment]
Thanks a lot @big-guy for your code review! It made me learn many things!

As far as I understand, `@OutputFiles` is not supposed to be slower than `@OutputDirectory`, right?

1. Regarding the managed properties, I am going to have a look at it. We currently support 5.6+.
2. Regarding the dumb tasks, I see what you mean. The philosophy behind is that the plugin is not just responsible of declaring tasks, it also configures them. Your example regarding `npmInstall` task illustrates that.
3. Ok, I am going to remove that.
4. By using [injection](https://docs.gradle.org/current/userguide/custom_gradle_types.html#service_injection)? If so, I see what you mean and it's quite easy to change that.

I created issue #100 in our project to apply your recommendations.

**Comment by li-wjohnson** [comment]
Can I add a related suggestion here? I'd like to see a `zip` that works with `Iterable<Provider<T>>`, or maybe some additions to `HasMultipleValues`.

We have a ListProperty, and we want to map each item to a Provider<T>. Right now we have (in kotlin):
```
listProperty.map { list -> list.map { it.returnsAProvider().get() }}
```
I'm not sure how to eliminate the `get` here, without using some crazy fold using `zip`. It would be nice to have something like:
```
listProperty.flatMapItems { item -> item.returnsAProvider() }
```
or
```
listProperty.flatMap { list -> providers.zip(list.map { it.returnsAProvider() }) }  // returns Provider<Iterable<T>>
```

**Comment by wolfs** [comment]
There now is `Provider.zip()`, I suppose this fixes the use-case. Please comment if it doesn't.
